lib/libuzfs/libuzfs_attr.c

```diff
@@ -0,0 +1,600 @@
+			err = nvlist_size(nvl, xattr_sa_size, NV_ENCODE_XDR);
+		}
+		if (err != 0) {
+			return (err);
+		}
+		return (EFBIG);
+	}
+
+	return (0);
+}
+
+static int
+libuzfs_inode_lookup_kvattr_zap(sa_handle_t *sa_hdl, sa_attr_type_t *sa_tbl,
+    const char *name, uint64_t *xattr_zap_obj)
+{
+	objset_t *os = sa_hdl->sa_os;
+	int err = sa_lookup(sa_hdl, sa_tbl[UZFS_ZXATTR],
+	    xattr_zap_obj, sizeof (*xattr_zap_obj));
+	if (err != ENOENT && err != 0) {
+		return (err);
+	}
+	if (err == 0) {
+		err = zap_contains(os, *xattr_zap_obj, name);
+	}
+	return (err);
+}
+
+int
+libuzfs_inode_set_kvattr(libuzfs_dataset_handle_t *dhp, uint64_t ino,
+    const char *name, const char *value, uint64_t size,
+    int flags, uint64_t *txg)
+{
+	// only support value size <= 8K
+	ASSERT3U(size, <=, ZAP_MAXVALUELEN);
+
+	sa_handle_t *sa_hdl = NULL;
+	objset_t *os = dhp->os;
+	int err = sa_handle_get(os, ino, NULL, SA_HDL_PRIVATE, &sa_hdl);
+	if (err != 0) {
+		return (err);
+	}
+
+	nvlist_t *nvl;
+	sa_attr_type_t *sa_tbl = dhp->uzfs_attr_table;
+	err = libuzfs_get_nvlist_from_handle(sa_tbl, &nvl, sa_hdl);
+	if (err == ENOENT) {
+		err = nvlist_alloc(&nvl, NV_UNIQUE_NAME, KM_SLEEP);
+	}
+	if (err != 0) {
+		goto out_handle;
+	}
+
+	boolean_t existed_in_sa = nvlist_exists(nvl, name);
+	boolean_t sa_space_enough = B_TRUE;
+	size_t xattr_sa_size = 0;
+	err = libuzfs_kvattr_set_nvlist(nvl, name, value, size, &xattr_sa_size);
+	if (err == EFBIG) {
+		sa_space_enough = B_FALSE;
+	} else if (err != 0) {
+		goto out_nvl;
+	}
+
+	char *xattr_sa_data = NULL;
+	if ((existed_in_sa && xattr_sa_size > 0) || sa_space_enough) {
+		xattr_sa_data = vmem_alloc(xattr_sa_size, KM_SLEEP);
+		err = nvlist_pack(nvl, &xattr_sa_data, &xattr_sa_size,
+			NV_ENCODE_XDR, KM_SLEEP);
+		if (err != 0) {
+			goto out_free_sa_data;
+		}
+	}
+
+	boolean_t existed_in_zap = B_FALSE;
+	uint64_t xattr_zap_obj = DMU_NEW_OBJECT;
+	uint64_t zap_entries_count = 0;
+	if (!existed_in_sa || !sa_space_enough) {
+		err = libuzfs_inode_lookup_kvattr_zap(sa_hdl,
+		    sa_tbl, name, &xattr_zap_obj);
+		existed_in_zap = err == 0;
+		ASSERT(!existed_in_sa || !existed_in_zap);
+		if (existed_in_zap && !sa_space_enough) {
+			err = zap_count(os, xattr_zap_obj, &zap_entries_count);
+		}
+
+		if (err != 0 && err != ENOENT) {
+			goto out_free_sa_data;
+		}
+	}
+
+	dmu_tx_t *tx = dmu_tx_create(os);
+	dmu_tx_hold_sa(tx, sa_hdl, B_TRUE);
+	if (existed_in_zap || !sa_space_enough) {
+		if (existed_in_zap && sa_space_enough &&
+		    zap_entries_count == 1) {
+			dmu_tx_hold_free(tx, xattr_zap_obj, 0, DMU_OBJECT_END);
+		} else {
+			dmu_tx_hold_zap(tx, xattr_zap_obj,
+			    !sa_space_enough, name);
+		}
+	}
+
+	if ((err = dmu_tx_assign(tx, TXG_WAIT)) == 0) {
+		if (existed_in_sa || sa_space_enough) {
+			if (xattr_sa_size == 0) {
+				VERIFY0(sa_remove(sa_hdl,
+				    sa_tbl[UZFS_XATTR], tx));
+			} else {
+				VERIFY0(sa_update(sa_hdl, sa_tbl[UZFS_XATTR],
+				    xattr_sa_data, xattr_sa_size, tx));
+			}
+		}
+
```

Comment:
sa_space_enough and xattr_sa_size == 0 can't both be true. so it's better to refine this check:

if (sa_space_enough) {
    // sa_update()
} else if (existed_in_sa && xattr_sa_size == 0) {
    // this is a corner case
    // sa_remove()
}

Reply:
if existed_in_sa && xattr_sa_size != 0 && ! sa_space_enough, we still need to call sa_update. it's better to modify like this:

```
if (existed_in_sa && xattr_sa_size == 0) {
	sa_remove();
} else if (existed_in_sa || sa_space_enough) {
	sa_update();
}
```

Reply:
make sense